openwisp_controller/config/api/serializers.py

```diff
@@ -0,0 +1,239 @@
+from django.db import transaction
+from django.db.models import Q
+from django.utils.translation import ugettext_lazy as _
+from rest_framework import serializers
+from swapper import load_model
+
+from openwisp_users.api.mixins import FilterSerializerByOrgManaged
+from openwisp_utils.api.serializers import ValidatedModelSerializer
+
+Template = load_model('config', 'Template')
+Vpn = load_model('config', 'Vpn')
+Device = load_model('config', 'Device')
+Config = load_model('config', 'Config')
+Organization = load_model('openwisp_users', 'Organization')
+
+
+class BaseMeta:
+    read_only_fields = ['created', 'modified']
+
+
+class BaseSerializer(FilterSerializerByOrgManaged, ValidatedModelSerializer):
+    pass
+
+
+class TemplateSerializer(BaseSerializer):
+    config = serializers.JSONField(initial={})
+    tags = serializers.StringRelatedField(many=True, read_only=True)
+    default_values = serializers.JSONField(required=False, initial={})
+
+    class Meta(BaseMeta):
+        model = Template
+        fields = [
+            'id',
+            'name',
+            'organization',
+            'type',
+            'backend',
+            'vpn',
+            'tags',
+            'default',
+            'required',
+            'default_values',
+            'config',
+            'created',
+            'modified',
+        ]
+
+    def validate_vpn(self, value):
+        """
+        Ensure that VPN can't be added when
+        template `Type` is set to `Generic`.
+        """
+        if self.context['request'].data['type'] == 'generic':
+            raise serializers.ValidationError(
+                _("To select a VPN, set the template type to 'VPN-client'")
+            )
+        return value
+
+
+class VpnSerializer(BaseSerializer):
+    config = serializers.JSONField(initial={})
+
+    class Meta(BaseMeta):
+        model = Vpn
+        fields = [
+            'id',
+            'name',
+            'host',
+            'organization',
+            'key',
+            'ca',
+            'cert',
+            'backend',
+            'notes',
+            'dh',
+            'config',
+            'created',
+            'modified',
+        ]
+
+
+class FilterTemplatesByOrganization(serializers.PrimaryKeyRelatedField):
+    def get_queryset(self):
+        user = self.context['request'].user
+        if user.is_superuser:
+            queryset = Template.objects.all()
+        else:
+            userOrgList = user.organizations_dict.keys()
+            query = Q(organization=None)
+            for org in userOrgList:
+                query |= Q(organization=org)
+            queryset = Template.objects.filter(query)
```

Comment:
I can't find the previous thread on this code, but bottom line there was that there is a problem with this code as well,
Consider if user is in 2 organizations (org1 & org2) and they are creating device in org2.
We want to show them templates from org2 & `organization=None` filter.

Makes sense?
This is also wrong according to that conversation because this will filter for both org1 and org2, while the code before this would have filtered for org1. 

See the problem?

Reply:
Yes, I understand but I am not sure if this can be done with DRF. :thinking: , but still this approach is better then the previous one, and this will not cause any error, because if a user selects org2 as organization of the device, and then selects any templates belonging to org1, then a validation error is raised not allowing the user to create the instance. 

Also, If there is anyway we can achieve it? I would love to get any suggestion/material/reference for achieving this. :smile:

Reply:
hey guys, I really do not follow you and I don't understand why this is being done, is it being done for read operations or write operations?

For read, all the templates assigned to a device must be shown to any user.
For write, we should allow only templates that are either shared or belong to one of the organizations the user manages (not all organizations like you're doing here Manish).
I think we should do a change like the one I propose here: https://github.com/openwisp/openwisp-users/issues/247#issuecomment-831644780 and use that.

@ManishShah120 if you want to test such a solution in this module you can do so, but I avoid to rewrite the whole serializer like you're doing here because that's duplication. I would inherit `FilterByOrganizationManaged` and override the `get_organization_queryset` method.

Reply:
PS: watch out from camel case names like `userOrgList`, pelase keep the naming convention consistent.

Reply:
@nemesisdesign  
https://github.com/openwisp/openwisp-users/issues/247#issuecomment-831644780
Will return the list of all the orgs that user manages, how do we filter which one we want to use? For that, I proposed, let's just the organization selected for the device and we rely on device's organization not user's organization for filtering the templates on device creation page.

Reply:
@atb00ker Suppose you want to create a device, with the API and you belong to `org1` & `org2`, then since the device has not been created, how would the `templates` be filtered?. So here you suggested something like when a user selects the organization of a device after that the templates belonging to that org should be populated right, but this is not possible with DRF.

So, What I am doing here is:-
I am filtering the templates based on the organization of the user, from that the user can select the organization he want to create a device from and then select the corresponding templates belonging to the organization of the device, and if the user selects templates not belonging to the org of the device, then a validation error is raised.

Reply:
> hey guys, I really do not follow you and I don't understand why this is being done, is it being done for read operations or write operations?

This function filters out the templates based on the organization of the user.

Reply:
@ManishShah120 ok let's find a solution that can work also in nested serializers then, is that an issue?
@atb00ker I would keep it simple and avoid further filtering, we have done that in the django admin and is not trivial, it requires a lot of work to be done correctly, this is more a frontend thing, we expect software which will consume this API to do that work of filtering out the right values to show to users, don't you think?

Reply:
@nemesisdesign  
https://github.com/openwisp/openwisp-users/issues/247#issuecomment-831644780
Will return the list of all the orgs that user manages, how do we filter which one we want to use? For that, I proposed, let's just the organization selected for the device and we rely on device's organization not user's organization for filtering the templates on device creation page.

Reply:
@ManishShah120 Had a conversation with Fed, 
we will be going for: 
```
we show all templates from
- org=None
- any org were user is manager
 ```
 
 Because the case I mentioned is a corner case and the problem wouldn't be seen in normal usage as extended apps can take care of that.. if we need it, we will come to it in future! :smile: